lab_2_tokenize_by_bpe/main.py

```diff
@@ -58,6 +105,30 @@ def train(
     :param num_merges: required number of new tokens
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(num_merges, int):
+        return None
+    dict_of_token_pairs = count_tokens_pairs(word_frequencies)
+    if dict_of_token_pairs is None:
+        return None
+    list_of_max_token_pairs = []
+    for i in range(0, num_merges):
+        if len(list_of_max_token_pairs) == num_merges:
+            break
+        temp_list_of_max_token_pairs = []
+        max_token_pair = max(dict_of_token_pairs.values())
+        for token_pair in dict_of_token_pairs:
+            if dict_of_token_pairs[token_pair] == max_token_pair:
+                temp_list_of_max_token_pairs.append(token_pair)
+        temp_list_of_max_token_pairs.sort(key=lambda x: (-len(x), x))
+        for temp_token_pair in temp_list_of_max_token_pairs:
+            if len(list_of_max_token_pairs) < num_merges:
+                list_of_max_token_pairs.append(temp_token_pair)
+    new_dict = word_frequencies
+    for new_token_pair in list_of_max_token_pairs:
+        new_dict = merge_tokens(new_dict, new_token_pair)
+    if new_dict is None:
+        return None
+    return new_dict
```

Comment:
The Unittests continue to fail but I cannot understand the error

Reply:
Here is how you should approach the situation:

1. Focus on one failing test at a time
2. Understand what this test is about. Usually the naming is self-explanatory.
3. Find the input value that the test uses and the result value it expects. Decide if you agree with this expected value, whether it is indeed the correct result and why. Compare it to the return value your function outputs. This should give you an idea what could be wrong with your implementation. You can use debugging mode in pycharm, or print values at each step to track the behaviour of your function. 

If you continue to struggle, try to formulate the specific question and ask it in chat.